c/tskit/tables.c

```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+    tsk_size_t metadata_length;
+    tsk_size_t derived_state_length;
+    tsk_id_t s, n, p;
+    if (row >= (tsk_id_t) source->num_rows) {
+        ret = TSK_ERR_MUTATION_OUT_OF_BOUNDS;
+        goto out;
+    }
+    s = source->site[row];
+    if (s >= 0) {
+        s = site_map[s];
+    }
+    n = source->node[row];
+    if (n >= 0) {
+        n = node_map[n];
+    }
+    p = source->parent[row];
+    if (p >= 0) {
+        p = mutation_map[p];
+    }
+    metadata_length = source->metadata_offset[row + 1] - source->metadata_offset[row];
+    derived_state_length
+        = source->derived_state_offset[row + 1] - source->derived_state_offset[row];
+    ret = tsk_mutation_table_add_row(dest, s, n, p,
+        source->derived_state + source->derived_state_offset[row], derived_state_length,
+        source->metadata + source->metadata_offset[row], metadata_length);
+    if (ret < 0) {
+        goto out;
+    }
+out:
+    return ret;
+}
+
+static int TSK_WARN_UNUSED
+site_table_copy_row(tsk_site_table_t *source, tsk_site_table_t *dest, tsk_id_t row)
+{
+    int ret = 0;
+    tsk_size_t metadata_length;
+    tsk_size_t ancestral_state_length;
+    if (row >= (tsk_id_t) source->num_rows) {
+        ret = TSK_ERR_SITE_OUT_OF_BOUNDS;
+        goto out;
+    }
+    metadata_length = source->metadata_offset[row + 1] - source->metadata_offset[row];
+    ancestral_state_length
+        = source->ancestral_state_offset[row + 1] - source->ancestral_state_offset[row];
+    ret = tsk_site_table_add_row(dest, source->position[row],
+        source->ancestral_state + source->ancestral_state_offset[row],
+        ancestral_state_length, source->metadata + source->metadata_offset[row],
+        metadata_length);
+    if (ret < 0) {
+        goto out;
+    }
+out:
+    return ret;
+}
+
+static int TSK_WARN_UNUSED
+migration_table_copy_row(tsk_migration_table_t *source, tsk_migration_table_t *dest,
+    tsk_id_t row, tsk_id_t *node_map, tsk_id_t *population_map)
+{
+    int ret = 0;
+    tsk_size_t metadata_length;
+    if (row >= (tsk_id_t) source->num_rows) {
+        ret = TSK_ERR_MIGRATION_OUT_OF_BOUNDS;
+        goto out;
+    }
+    metadata_length = source->metadata_offset[row + 1] - source->metadata_offset[row];
+    ret = tsk_migration_table_add_row(dest, source->left[row], source->right[row],
+        node_map[source->node[row]], population_map[source->source[row]],
+        population_map[source->dest[row]], source->time[row],
+        source->metadata + source->metadata_offset[row], metadata_length);
+    if (ret < 0) {
+        goto out;
+    }
+out:
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_table_collection_subset(
+    tsk_table_collection_t *self, tsk_id_t *nodes, size_t num_nodes)
+{
+    int ret = 0;
+    tsk_id_t k, n, m, ind, pop, mut, site;
+    tsk_id_t *node_map, *individual_map, *population_map, *site_map, *mutation_map;
+    tsk_table_collection_t tables;
+
+    node_map = calloc(self->nodes.num_rows, sizeof(tsk_id_t));
```

Comment:
Since we're memsetting all of these straight afterwards, might as well use ``malloc`` here.

Reply:
Also, better do use sizeof(*node_map), e.g.
```c
node_map = malloc(self->nodes.num_rows * sizeof(*node_map));
```
this is a bit more robust and less error prone.